Fix exponent padding when deriving rsapubkey to ssh

peptekmail opened the pull request:
Padding is sometimes needed if a nonstandard exponent is chosen.

The fix is just a couple of lines in cert.c

But the integration-test requires a certificate to be pushed to LDAP and the output should match the pubkey derived from the original certificate via p11-tool and openssh.

peptekmail commented:
This problem probably arised when switching from nss to openssl.
Backporting to redhat8 would be nice.
Debian/Ubuntu is about to switch to openssl and will probably notice the problem if they use the feature.

peptekmail commented:
@sumit-bose 
This fix solves one of several problems with some 3:rd party certificates, this problem is not isolated to rsassa-pss but could occur om any rsa-based client-cert.
However most CA:s go for the default openssl settings to avoid this problem.
I have not found any similair code anywhere else in SSSD but someone more experienced might know better.

sumit-bose commented:
Hi,

thanks for the patch, I agree with the fix. You are right, this issue is not specific to rsassa-pss but is related to how BIGNUM handles negative numbers (highest bit must be set). The similar code is just in the following lines in `rsa_pub_key_to_ssh()` where the `0` is added unconditionally to the modulus. Iirc in the old NSS based code there was a similar if-block than the one you used here. I guess adding the `0` unconditionally to the exponent should work as well.

I added some inline comments as well.

bye,
Sumit

sumit-bose commented on src/tests/test_CA/SSSD_test_cert_0005.config:
```diff
@@ -19,3 +19,4 @@ keyUsage = critical, nonRepudiation, digitalSignature, keyEncipherment
 extendedKeyUsage = clientAuth
 subjectAltName = email:sssd-devel@lists.fedorahosted.org,URI:https://github.com/SSSD/sssd//
 authorityInfoAccess = OCSP;URI:http://ocsp.my.server.test/
+
```
Can you remove this empty last line?

peptekmail replied:
Done

sumit-bose commented on src/tests/test_CA/Makefile.am:
```diff
@@ -118,6 +119,15 @@ softhsm2_one.conf:
 	@echo "objectstore.backend = file" >> $@
 	@echo "slots.removable = true" >> $@
 
+#Export cert from softhsm2 via p11tool, should produce the same as openssl
```
I do not understand this comment. You use the *.der files to add the certificate to LDAP, how is this related to openssl?

peptekmail replied:
Initially I chased differences between openssl and gnutls(openssl ignorerd some settings while writing and reading certs that made everything produced by openssl work perfect)
I can remove half the comment since that problem solved itself.

sumit-bose commented on src/tests/intg/test_ssh_pubkey.py:
```diff
@@ -261,12 +283,40 @@ def add_user_with_many_keys(request, ldap_conn):
+@pytest.fixture
+def add_user_with_ssh_cert(request, ldap_conn):
+    # Add a certificate to ldap, to manually test a cert from a smartcard.
+    # Export it and insert below.
```
What do you mean with the second line of the comment?

peptekmail replied:
Removed since it is not related tonthis problem.

sumit-bose commented on src/tests/intg/test_ssh_pubkey.py:
```diff
@@ -235,6 +241,22 @@ def test_ssh_pubkey_retrieve(add_user_with_ssh_key):
+def test_ssh_pubkey_retrieve_cert(add_user_with_ssh_cert):
+    """
+    Test that we can retrieve an SSH public key derived from a cert in ldap.
+    Compare with the sshpubkey derived via openssl, they should match.
```
The pubsshkey files are created by ssh-keygen and not by OpenSSL.

alexey-tikhonov commented:
Please, avoid `Merge branch ... ` commit (use rebase) and please squash patches where having separate commits doesn't make much sense.

peptekmail commented:
> Hi,
> 
> thanks for the patch, I agree with the fix. You are right, this issue is not specific to rsassa-pss but is related to how BIGNUM handles negative numbers (highest bit must be set). The similar code is just in the following lines in `rsa_pub_key_to_ssh()` where the `0` is added unconditionally to the modulus. Iirc in the old NSS based code there was a similar if-block than the one you used here. I guess adding the `0` unconditionally to the exponent should work as well.
> 
> I added some inline comments as well.
> 
> bye,
> Sumit

I tester. O

> Hi,
> 
> thanks for the patch, I agree with the fix. You are right, this issue is not specific to rsassa-pss but is related to how BIGNUM handles negative numbers (highest bit must be set). The similar code is just in the following lines in `rsa_pub_key_to_ssh()` where the `0` is added unconditionally to the modulus. Iirc in the old NSS based code there was a similar if-block than the one you used here. I guess adding the `0` unconditionally to the exponent should work as well.
> 
> I added some inline comments as well.
> 
> bye,
> Sumit

Thanks for pointing that out.
I tested both conditional an unconditional padding with '0' om both exponent and modulus but this setup is the only one that seems to work every time.
Openssh does this in a diffrent way I do not fullty grasp so if anyone gets into trouble with this change that is where to look.
But the integration test should catch that and make it easy to compare diffrent certs.

peptekmail commented:
> Please, avoid `Merge branch ... ` commit (use rebase) and please squash patches where having separate commits doesn't make much sense.

Hopefully everything is cleansed now. One commit should be enough.

sumit-bose commented:
Hi,

thanks for the fixes and squashing everything together. ACK.

bye,
Sumit

pbrezina commented:
Pushed PR: https://github.com/SSSD/sssd/pull/5566

* `master`
    * 0e1452421fc99c094e1c23ced4efc44993e9702b - TEST: FIX: When generating a ssh pubkey from a cert extra padding is needed if a nonstandard eponent is chosen.
